src/zkscript/transaction_introspection/transaction_introspection.py

```diff
@@ -0,0 +1,220 @@
+
+from src.zkscript.types.stack_elements import StackBaseElement
+from src.zkscript.util.utility_scripts import (
+    bool_to_moving_function,
+    bytes_to_unsigned,
+    int_sig_to_s_component,
+    move,
+    nums_to_script,
+    pick,
+    reverse_endianness,
+    roll,
+)
+
+pushtx_bit_shift_data = {
+    2: {
+        "signature_prefix": bytes.fromhex("3045022100"),
+        # R = 2^2 * G
+        "R": bytes.fromhex("02e493dbf1c10d80f3581e4904930b1404cc6c13900ee0758474fa94abe8c4cd13"),
+        # P = a * G s.t. a * R_x = -1 mod GROUP_ORDER_INT
+        "P": bytes.fromhex("034218426b38c75b706db9010aad7795fd05b872060921c048d9a679d8878c7660"),
+    },
+    3: {
+        "signature_prefix": bytes.fromhex("30440220"),
+        # R = 2^3 * G
+        "R": bytes.fromhex("022f01e5e15cca351daff3843fb70f3c2f0a1bdd05e5af888a67784ef3e10a2a01"),
+        # P = a * G s.t. a * R_x = -1 mod GROUP_ORDER_INT
+        "P": bytes.fromhex("03ad36fad55727ebf76f8af96c7c2df9a298dc21d6c15269fdedfd47a70b327637"),
+    },
+}
+
+
+class TransactionIntrospection:
+    """Class generating Bitcoin scripts that achieve transaction introspection."""
+
+    @staticmethod
+    def pushtx(
+        sighash_value: SIGHASH,
+        sig_hash_preimage: StackBaseElement = StackBaseElement(0),  # noqa: B008
+        rolling_option: bool = True,
+        clean_constants: bool = True,
+        verify_constants: bool = True,
+        is_checksigverify: bool = True,
+        is_opcodeseparator: bool = False,
+    ) -> Script:
+        """Construct PUSHTX locking script.
+
+        Stack input:
+            - stack:    [GROUP_ORDER_INT, Gx, Gx_bytes, .., sig_hash_preimage, ..]
+            - altstack: []
+        Stack output:
+            - stack:    [GROUP_ORDER_INT, Gx, Gx_bytes, .., sig_hash_preimage, ..] or fail
+            - altstack: []
+
+        Args:
+            sighash_value (SIGHASH): Sighash flag with which the message should be constructed
+            sig_hash_preimage (StackBaseElement): Position in the stack of the sig_hash_preimage.
+                Defaults to StackBaseElement(0) (on top of the stack).
+            rolling_option (int): Whether to roll or pick sig_hash_preimage. Defaults to `True` (roll).
+            clean_constants (bool): Whether to clean the constants needed for the script:
+                GROUP_ORDER_INT, Gx, Gx_bytes. Defaults to `True`.
+            verify_constants (bool): Whether or not to verify the constants used for the script.
+            is_checksigverify (bool): Whether to execute OP_CHECKSIGVERIFY or OP_CHECKSIG. Defaults
+                to `True` (OP_CHECKSIGVERIFY).
+            is_opcodeseparator (bool): Whether to prepend the signature verification with an
+                OP_CODESEPARATOR (so that the code of pushtx is not signed). Defaults to `False`.
+
+        Returns:
+            The Bitcoin Script of PUSHTX, which fails unless sig_hash_preimage is the message digest of the
```

Comment:
Here I would just say the PUSHTX locking script. All the rest can be in a note part of the doctring